src/mongocxx/test/versioned_api.cpp

```diff
@@ -0,0 +1,200 @@
+    // Store it in the set of client options
+    const auto client_opts =
+        options::client{}
+            .server_api_opts(server_api_opts); // Set the version and options
+    // Create a new client with the options
+    mongocxx::client client{client_uri, client_opts};
+    // End Versioned API Example 3
+}
+
+TEST_CASE("Versioned API, non-strict with deprecation errors") {
+    instance::current();
+    if (!is_server_v5_or_newer()) {
+        return;
+    }
+    // Start Versioned API Example 4
+    using namespace mongocxx;
+    uri client_uri{"mongodb://localhost"};
+    // Create an option set for API v1
+    const auto server_api_opts =
+        options::server_api{options::server_api::version_from_string("1")}
+            .deprecation_errors(true);  // Enable deprecation errors
+    // Store it in the set of client options
+    const auto client_opts =
+        options::client{}
+            .server_api_opts(server_api_opts); // Set the version and options
+    // Create a new client with the options
+    mongocxx::client client{client_uri, client_opts};
+    // End Versioned API Example 4
+}
+// clang-format on
+
+TEST_CASE("Versioned API, with insert-many for 'count' migration") {
+    instance::current();
+    if (!is_server_v5_or_newer()) {
+        return;
+    }
+
+    using namespace mongocxx;
+    mongocxx::client client{
+        uri{},
+        options::client{}.server_api_opts(
+            options::server_api{options::server_api::version::k_version_1}.strict(true))};
+    using namespace bsoncxx::builder::basic;
+
+    // Drop in case we have stale data
+    client.database("db").collection("sales").drop();
+
+    // clang-format off
+    // Start Versioned API Example 5
+    auto db = client.database("db");
+    auto sales = db.collection("sales");
+    auto make_sale =
+        [](int id, std::string item_name, int price, int quanitity, std::string time) {
+            return make_document(kvp("_id", id),
+                                 kvp("item", item_name),
+                                 kvp("price", price),
+                                 kvp("quantity", quanitity),
+                                 kvp("date", time));
+        };
+    auto items = {
+        //       _id  item    price  qty  date
+        make_sale(1,  "abc",  10,    2,   "2021-01-01T08:00:00Z"),
+        make_sale(2,  "jkl",  20,    1,   "2021-02-03T09:00:00Z"),
+        make_sale(3,  "xyz",  5,     5,   "2021-02-03T09:05:00Z"),
+        make_sale(4,  "abc",  10,    10,  "2021-02-15T08:00:00Z"),
+        make_sale(5,  "xyz",  5,     10,  "2021-02-15T09:05:00Z"),
+        make_sale(6,  "xyz",  5,     5,   "2021-02-15T12:05:10Z"),
+        make_sale(7,  "xyz",  5,     10,  "2021-02-15T14:12:12Z"),
+        make_sale(8,  "abc",  10,    5,   "2021-03-16T20:20:13Z"),
+    };
+    sales.insert_many(items);
+    // End Versioned API Example 5
+    // clang-format on
+
+    try {
+        db.run_command(make_document(kvp("count", "sales")));
```

Comment:
I think this is causing a test failure on the `test_versioned_api` variant.

That variant is described in the Versioned API test readme: https://github.com/mongodb/specifications/tree/master/source/versioned-api/tests#testing-with-required-api-version

I believe the cause is:
- The run_command helper does not append server API parameters. This is specified in the Versioned API section [Generic command helper](https://github.com/mongodb/specifications/blob/df6be82f865e9b72444488fd62ae1eb5fca18569/source/versioned-api/versioned-api.rst#generic-command-helper).
- The `test_versioned_api` task configures the server and requires a server API version in all commands. That is configured in the [Evergreen config](https://github.com/mongodb/mongo-cxx-driver/blob/2fdfd62c60370f660c935f82c1f67b7d4b26d9ab/.mci.yml#L806:L806).

I think the solution is to use the deprecated `count` helper. That should append a server API. To avoid the deprecation warnings from failing compile, wrap the call in `BSONCXX_SUPPRESS_DEPRECATION_WARNINGS_BEGIN` and `BSONCXX_SUPPRESS_DEPRECATION_WARNINGS_END`.

This also alerted me that my corresponding change in CDRIVER-4093 did not take this into account, and also used the generic C driver command helper. I moved CDRIVER-4093 back to "In Progress" to address that.

Reply:
What is the deprecated `count` helper that you refer to? I can't seem to find anything that is deprecated with a "count" name.

Reply:
Ah, my apologies, I do not see one either. Some drivers have a deprecated `count` helper, which is distinct from `countDocuments` and `estimatedDocumentCount`. I assumed the C++ driver had one as part of the `mongocxx::collection` class.

For reference the [CRUD spec](https://github.com/mongodb/specifications/blob/df6be82f865e9b72444488fd62ae1eb5fca18569/source/crud/crud.rst#read) describes which underlying MongoDB command is sent for each of the three count helpers.

> The run_command helper does not append server API parameters. This is specified in the Versioned API section Generic command helper.

My understanding was incorrect. The generic command helpers do append API version parameters. After inspecting test failures introduced in [CDRIVER-4093](https://jira.mongodb.org/browse/CDRIVER-4093), I see that libmongoc's `mongoc_database_command_with_opts` applies the version API as expected.

I think there is nothing to change here. I suspect the `The apiVersion parameter is required, please configure your MongoClient's API` error from [the `test_version_api` task](https://spruce.mongodb.com/task/cxx_driver_versioned_api__os~windows_2k8_mongodb_version~latest_test_versioned_api_patch_c2932becd7b69dfa15bfc6a8cda6038dfc33a600_6114951457e85a7025bb1f47_21_08_12_03_27_17/logs?execution=0) may be due to the `is_server_v5_or_newer` check above.